Refactor graphic option callbacks

AJenbo opened the pull request:
Fixes #3697 #3791

Only the relevant parts are reset when changing graphics options.

- Turning fullscreen on in settings doesn't flicker on Mac
- `-x` has been removed
- `alt+enter` changes the setting instead of being temporary
- Allow systems with no GPU to use CPU based rendering (this was an issue for [SerenityOS](https://github.com/SerenityOS/serenity) when @awesomekling [ported DeviutionX](https://www.youtube.com/watch?v=ZOzZ8R4gphE) to it, despite the CPU being powerful enough). Even if the CPU isn't powerful enough it should allow users to switch off upscalling in settings for better performance, so win win.
- When exiting fullscreen mode the window is correctly resized instead of being semi maximized
- Flipping interger scalling no longer break input on iOS (#3788)
- Remove stripped transparancy
- #3791
- Fixes window not updating after resizing on linux

StephenCWills reviewed (changes requested):
Something is definitely not right with integer scaling. It doesn't really seem like it was working correctly before this PR either, but with this PR I'm able to cause graphical glitches and crashes with certain combinations of integer scaling and fullscreen changes that weren't happening before. For instance, the following process never seems to fail to cause such issues.

1. Start game in fullscreen
2. Enter settings menu and change Integer Scaling
3. Exit the settings menu and enter a new single player game
4. Use Alt+Enter to toggle fullscreen

StephenCWills commented on Source/utils/display.cpp:
```diff
@@ -278,16 +300,13 @@ void ReinitializeRenderer()
 	Size windowSize = { current.current_w, current.current_h };
 	AdjustToScreenGeometry(windowSize);
 #else
-	if (texture)
-		texture.reset();
```
Better add these back. I got a failed assertion when changing the resolution in the settings menu. It may seem redundant to do it both here and in `ReinitializeTexture()`, but the texture does need to be destroyed before the renderer to avoid errors when we inevitably attempt to destroy the texture later on.

StephenCWills replied:
I slipped this in with the integer scaling fixes.

StephenCWills commented:
This is pretty minor, but using Alt+Enter while in the settings menu doesn't update the value displayed next to the `Fullscreen` setting.

obligaron commented on Source/options.cpp:
```diff
@@ -792,14 +784,14 @@ GraphicsOptions::GraphicsOptions()
-	scaleQuality.SetValueChangedCallback(ReinitializeRenderer);
-	integerScaling.SetValueChangedCallback(ResizeWindow);
+	scaleQuality.SetValueChangedCallback(ReinitializeTexture);
+	integerScaling.SetValueChangedCallback(ReinitializeIntegerScale);
```
Doesn't integer scaling also affect the potential window size?
(see `CalculatePreferdWindowSize`/ `ResizeWindow` / `SetFullscreenMode`)

StephenCWills replied:
You hit the nail on the head. The changes to window size are actually the cause of the problem here. Just take them out and integer scaling works entirely as expected for me. Also, `dx_resize()` is no longer necessary.

StephenCWills replied:
@AJenbo FYI, it looks to me like `CalculatePreferdWindowSize()` was trying to adjust the resolution to the largest integer multiple of the requested resolution. However, that changes the game's resolution by updating `gnScreenWidth` and `gnScreenHeight` to that integer multiple. I think instead the desired effect is to set the logical size of the renderer to the requested resolution and the window size to the integer multiple in order to get the maximum scale possible for that display mode.

If you like, I can work on a commit for this PR to fix everything that came up based on my testing, and then you can rebase however you like.

AJenbo replied:
Thanks that would be appreciated, haven't been able to do much at all today.

StephenCWills replied:
Alright, I fixed all the issues I could find. I'm submitting my approval since AJenbo can't do it himself, but it would be good if someone else reviews my changes before merging.

StephenCWills commented on Source/utils/display.cpp:
```diff
@@ -337,6 +350,34 @@ void ReinitializeRenderer()
+	}
+#endif
+	force_redraw = 255;
+}
```
I don't think so. Rendering coordinates should only be affected by changes to the logical size. It's easy enough to test, though, so I'll double-check.

StephenCWills replied:
Yep, I've confirmed that so long as the logical size of the renderer does not change, we don't need to call `InitializeVirtualGamepad()`.

StephenCWills commented on Source/utils/display.cpp:
```diff
@@ -268,6 +269,27 @@ bool SpawnWindow(const char *lpWindowName)
+	if (SDL_RenderSetIntegerScale(renderer, *sgOptions.Graphics.integerScaling ? SDL_TRUE : SDL_FALSE) < 0) {
+		ErrSdl();
+	}
+	dx_resize();
```
The latest commit merely updates the window size without updating the render logical size. Since the logical size doesn't change, there is no need to reinitialize the texture or the global back buffer.

StephenCWills commented:
I thought about this a bit more, and I realize now that the integer scaling logic that updated the game's resolution was probably put in place to eliminate the black borders around the viewport. It actually makes some sense to do it this way when Fit to Screen is on, but we'll have to call `ResizeWindow()` if we intend to change the resolution like that.

AJenbo commented:
oh yeah, forgot about that.

StephenCWills commented:
Okay, that was actually significantly less complicated. 😅

AJenbo commented:
Alright went over the changes and fixed two more issues that we introduced :)
- crash when switching integer scaling if upscale is off
- switching upscaling didn't flip resiable window on/off

I think that if either of you are good with the latest changes then we can merge it :)
